### PR TITLE
Inital vhost implementation

### PR DESCRIPTION
**Pull request by arronwy**
This is a sync of the Cloud Hypervisor `dragonball` branch for the vhost crate.
It is a superset of PR #2 with additional fixes and enhancements coming from the Cloud Hypervisor project. This PR is an exact mirror of the current [vhost crate](https://github.com/cloud-hypervisor/vhost/tree/dragonball)  used by the Cloud Hypervisor vhost implementation.

Fixes: #8

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
+name: vhost
```
Comment (andreeaflorescu) [inline .github/workflows/main.yml]: For consistency, can we use rust-vmm-ci?

File: Cargo.toml [hunk 2]
```diff
@@ -3,6 +3,23 @@ name = "vhost"
...
+vhost-user-master = []
+vhost-user-slave = []
 
 [dependencies]
+bitflags = "1.1.0"
+libc = "0.2.67"
+vmm-sys-util = ">=0.3.1"
+
+[dependencies.vm-memory]
+git = "https://github.com/rust-vmm/vm-memory"
```
Comment (andreeaflorescu) [inline Cargo.toml]: `vm-memory` is published to crates.io, no need to use it from github.

File: Cargo.toml [hunk 3]
```diff
@@ -3,6 +3,23 @@ name = "vhost"
...
+optional = true
+
+[dev-dependencies]
+tempfile = "3.1.0"
```
Comment (andreeaflorescu) [inline Cargo.toml]: You don't need this dependency as `tempfile` abstractions are already defined in vmm-sys-util.

File: LICENSE-BSD-Chronium [hunk 4]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2017 The Chromium OS Authors. All rights reserved.
```
Comment (andreeaflorescu) [inline LICENSE-BSD-Chronium]: Typo in file name. Probably we want to say "Chromium".

File: LICENSE-MIT [hunk 5]
```diff
@@ -0,0 +1,24 @@
+Permission is hereby granted, free of charge, to any
```
Comment (andreeaflorescu) [inline LICENSE-MIT]: Didn't we drop this license in favor of BSD-3-Clause?

File: .github/workflows/main.yml (hunk 1 above)
Comment (arronwy) [reply .github/workflows/main.yml]: sure, we switch to use rust-vmm-ci

File: Cargo.toml (hunk 2 above)
Comment (arronwy) [reply Cargo.toml]: ok, we will update this part

File: Cargo.toml (hunk 3 above)
Comment (arronwy) [reply Cargo.toml]: Yes, we also plan to remove the depdency for tempfile

File: LICENSE-BSD-Chronium (hunk 4 above)
Comment (arronwy) [reply LICENSE-BSD-Chronium]: Yes, will rename this file

File: LICENSE-MIT (hunk 5 above)
Comment (arronwy) [reply LICENSE-MIT]: Yes, I will remove this licence

**Comment by sameo** [comment]
@arronwy There are 2 reverts on that PR: https://github.com/rust-vmm/vhost/pull/9/commits/cf857aab86e229eb19f10eab676a7954aa3a2a2b and https://github.com/rust-vmm/vhost/pull/9/commits/c95ec2137d7571180bb7c083c369a0dbf28ff265

Could you please remove them and remove the patches that were reverted? This will make this long PR slightly simpler.

**Comment by arronwy** [comment]
> @arronwy There are 2 reverts on that PR: [cf857aa](https://github.com/rust-vmm/vhost/commit/cf857aab86e229eb19f10eab676a7954aa3a2a2b) and [c95ec21](https://github.com/rust-vmm/vhost/commit/c95ec2137d7571180bb7c083c369a0dbf28ff265)
> 
> Could you please remove them and remove the patches that were reverted? This will make this long PR slightly simpler.

OK, I'll submit a PR to remove them

**Comment by arronwy** [comment]
> @arronwy There are 2 reverts on that PR: [cf857aa](https://github.com/rust-vmm/vhost/commit/cf857aab86e229eb19f10eab676a7954aa3a2a2b) and [c95ec21](https://github.com/rust-vmm/vhost/commit/c95ec2137d7571180bb7c083c369a0dbf28ff265)
> 
> Could you please remove them and remove the patches that were reverted? This will make this long PR slightly simpler.

Just submitted the PR to fix this issue: https://github.com/cloud-hypervisor/vhost/pull/14

**Comment by sameo** [comment]
> > @arronwy There are 2 reverts on that PR: [cf857aa](https://github.com/rust-vmm/vhost/commit/cf857aab86e229eb19f10eab676a7954aa3a2a2b) and [c95ec21](https://github.com/rust-vmm/vhost/commit/c95ec2137d7571180bb7c083c369a0dbf28ff265)
> > Could you please remove them and remove the patches that were reverted? This will make this long PR slightly simpler.
> 
> OK, I'll submit a PR to remove them

No, please git push on this PR with both reverts and the commits they reverts removed from the PR. Please let me know if you need further help with this.

**Comment by arronwy** [comment]
> > 
> > OK, I'll submit a PR to remove them
> 
> No, please `git push --force` on this PR with both reverts and the commits they reverts removed from the PR. Please let me know if you need further help with this.

This PR is submitted from cloud-hypervisor:dragonball branch, and I can not git push --force on that branch : (, then I submit a PR in cloud-hypervisor:dragonball branch, if merged, it can auto update this PR.

I not sure wheter I was wrong at the begining to submit from cloud-hypervisor:dragonball branch, Should I submit from a forked local branch instead of cloud-hypervisor:dragonball branch?

**Comment by andreeaflorescu** [comment]
> > > OK, I'll submit a PR to remove them
> > 
> > 
> > No, please `git push --force` on this PR with both reverts and the commits they reverts removed from the PR. Please let me know if you need further help with this.
> 
> This PR is submitted from cloud-hypervisor:dragonball branch, and I can not git push --force on that branch : (, then I submit a PR in cloud-hypervisor:dragonball branch, if merged, it can auto update this PR.
> 
> I not sure wheter I was wrong at the begining to submit from cloud-hypervisor:dragonball branch, Should I submit from a forked local branch instead of cloud-hypervisor:dragonball branch?

@arronwy please submit a PR from a fork of rust-vmm/vhost. In my opinion, the whole history of how this was developed in another fork is not useful here. It also makes the PR hard to review for people that didn't contribute to that fork.

**Comment by sameo** [comment]
> > > OK, I'll submit a PR to remove them
> > 
> > 
> > No, please `git push --force` on this PR with both reverts and the commits they reverts removed from the PR. Please let me know if you need further help with this.
> 
> This PR is submitted from cloud-hypervisor:dragonball branch, and I can not git push --force on that branch : (, then I submit a PR in cloud-hypervisor:dragonball branch, if merged, it can auto update this PR.
> 
> I not sure wheter I was wrong at the begining to submit from cloud-hypervisor:dragonball branch, Should I submit from a forked local branch instead of cloud-hypervisor:dragonball branch?

Oh yes, you should. I did not realize you were pushing directly from the Cloud Hypervisor branch.

Once this PR is merged, we'll move Cloud Hypervisor to use the rust-vmm crate, and both the dragonball and master branches from cloud-hypervisor/vhost will be updated.

**Comment by arronwy** [comment]
@sameo  @andreeaflorescu thanks for your help, I created a new PR #10  which is forked from rust-vmm vhost.

**Pull request closed without merging**